Adds coverage on order-associated attributes

filipefurtad0 opened the pull request:
#### What? Why?

Closes #11226.

Adds test coverage to [attribute list](https://docs.google.com/spreadsheets/d/1hOM6UL4mWeRCFLcDQ3fTkbhbUQ2WvIUCCA1IerDBtUA/edit?pli=1#gid=0).

<!-- Explain why this change is needed and the solution you propose.
     Provide context for others to understand it. -->

There are three possible scenarios, when an order attribute or related attribute is changed:
- a new invoice is generated
- the existing invoice is updated
- nothing happens

This PR adds unit level coverage to the attribute list and the expected scenarios.

#### What should we test?
<!-- List which features should be tested and how.
     This can be similar to the Steps to Reproduce in the issue.
     Also think of other parts of the app which could be affected
     by your change. -->

- Green build
- It would be interesting to test all of these scenarios at least once, manually. We might consider adapting some of these test cases to system test level.

#### Release notes

<!-- Please select one for your PR and delete the other. -->

Changelog Category (reviewers may add a label for the release notes):

- [ ] User facing changes
- [ ] API changes (V0, V1 or DFC)
- [x] Technical changes only
- [ ] Feature toggled

<!-- Choose a pull request title above which explains your change to a
     a user of the Open Food Network app. -->

The title of the pull request will be included in the release notes.


#### Dependencies
<!-- Does this PR depend on another one?
     Add the link or remove this section. -->



#### Documentation updates
<!-- Are there any wiki pages that need updating after merging this PR?
     List them here or remove this section. -->

filipefurtad0 commented on spec/services/order_invoice_comparator_spec.rb:
```diff
@@ -115,19 +115,241 @@
+              end
+            end
+          end
+        end
```
... but it also passes here, which is the `generate new invoice section`. Any idea why this is happening @abdellani ?

abdellani replied:
Yes, this is not supposed to happen.
I'll inspect the issue.

abdellani replied:
This is happening because `order.payment_total` changes from 0 to `0.4575e2` after running  `payment.complete`.

filipefurtad0 replied:
Discussed and fixed on our live chat, resolving :pray:

abdellani commented on spec/services/order_invoice_comparator_spec.rb:
```diff
@@ -115,19 +115,241 @@
+
+        context "order cycle changes" do
+          it "returns false" do
+            pending("WIP - can't seem to change the order cycle on an order")
```
I'm getting a validation failure.
```ruby
     ActiveRecord::RecordInvalid:
       Validation failed: Distributor or order cycle cannot supply the products in your cart
```

filipefurtad0 replied:
Discussed and removed on our live chat, resolving :pray:

filipefurtad0 commented:
Together with @abdellani we've agreed that all relevant attributes should be set as shared examples, and ran both under the `generate new invoice` and `update invoice`: each test should pass on one and only on one of these scenarios.

I'll take on this re-structuring, before adding more tests :+1:

filipefurtad0 commented:
Hey @abdellani ,
There are a few pending test cases which pass on both scenarios - I'll still need to investigate those. But I'm pinging you here for a an early review, for your impression on the refactoring approach:
- I've created a separate spec file, to which I've moved/refactored the tests as shared examples, from the existing test file
- After all the tests were moved, I've deleted the original spec, and kept the refactored version

Happy for your thoughts - thank you :pray:

filipefurtad0 commented:
I've updated the commit history, to display (only):
- the added tests
- the restructuring into shared examples
- the updated `payment_total` tests, pending the closure #11350.

I hope this makes reviewing easier. There are a few otherpending test cases which I need to investigate, whether the unexpected outcome is from the test or the code itself.

But for now, and as discussed on delivery circle - build is green - we should move this to code review.

dacook commented on spec/services/order_invoice_comparator_spec.rb:
```diff
@@ -115,19 +115,217 @@
+        context "adjustment changes" do
+          it "creating a new adjustment returns true" do
+            create(:adjustment, order_id: order.id)
+            order.reload
```
Wait, I just figured it out, I think this would work (also removing the need for `order.reload`)

```suggestion
            order.adjustments << create(:adjustment)
```

filipefurtad0 replied:
Nice catch, thanks @dacook . 
I noticed though, that although adjustments array was being populated with a new adjustment, it was being created without an `order_id`; this seemed to make the tests fail.

By adding `order:` to it, made the tests pass again, like so:
`order.adjustments << create(:adjustment, order:)`

dacook commented on spec/services/order_invoice_comparator_shared_examples_spec.rb:
```diff
@@ -0,0 +1,89 @@
+      create(:invoice,
+             order:,
+             data: invoice_data_generator.serialize_for_invoice)
+    }
```
I think this setup is duplicated and could be shared in the above block.

filipefurtad0 replied:
Indeed, it is! Moved it up a level.

dacook commented on spec/services/order_invoice_comparator_spec.rb:
```diff
@@ -2,387 +2,399 @@
+    end
+  end
+
+  context "if included is #{included_boolean}" do
```
It might be a bit clearer if the full attribute name is included here: 

```suggestion
  context "if included_in_price is #{included_boolean}" do
```

filipefurtad0 replied:
Renamed the description :ok:

dacook reviewed (approved):
Looks good, I like how the final result is easy to read (I imagine it's a similar list to the spreadsheet).

I think we can make the code a bit shorter (see comments), but it's otherwise all good. Consider it optional :)

Tip for next reviewer: the diffs were hard to read, so I found it helpful to start by looking at the final file first: https://github.com/openfoodfoundation/openfoodnetwork/blob/2004abc8bec22d6b4f2aa63506bc6874f744db3d/spec/services/order_invoice_comparator_spec.rb

abdellani reviewed (approved):
Hi @filipefurtad0 

It looks good to me.

I like how the tests are grouped together. 

Thank you.

I'll move to that payment total issue after I finish with the new version (V3) of invoices.

mkllnk reviewed (approved):
Great!

There's still a bit of code duplication but I think that it's okay. It enables you to set only some to pending.
